File: mcp-server/src/claude_trade/tools/feedback.py
"""フィードバックループ: シグナル評価・学習ツール"""

from datetime import datetime, timedelta

from sqlalchemy import select, func
from sqlmodel import col

from claude_trade.db import async_session
from claude_trade.models import Signal, SignalOutcome, Lesson, Order


async def evaluate_signal(
    signal_id: str,
    price_at_signal: float,
    price_at_eval: float,
    evaluation: str,
    pnl: float | None = None,
) -> dict:
    """シグナルの事後評価を記録する

    Args:
        signal_id: 評価するシグナルの ID
        price_at_signal: シグナル発行時の価格
        price_at_eval: 評価時の価格
        evaluation: 事後評価テキスト
        pnl: 実現損益 (取引があった場合)
    """
    async with async_session() as session:
        stmt = select(Signal).where(Signal.id == signal_id)
        result = await session.execute(stmt)
        signal = result.scalar_one_or_none()

        if not signal:
            return {"error": f"Signal {signal_id} not found"}

        price_change_pct = ((price_at_eval - price_at_signal) / price_at_signal * 100) if price_at_signal > 0 else 0.0

        # 方向が合っていたか判定
        if signal.signal_type == "buy":
            direction_correct = price_change_pct > 0
        elif signal.signal_type == "sell":
            direction_correct = price_change_pct < 0
        elif signal.signal_type == "avoid":
            direction_correct = price_change_pct < 0  # 避けて正解だったか
        else:  # hold
            direction_correct = abs(price_change_pct) < 2.0  # 大きく動かなかったか

        outcome = SignalOutcome(
            signal_id=signal.id,
            symbol=signal.symbol,
            signal_type=signal.signal_type,
            strength=signal.strength,
            confidence=signal.confidence,
            source_strategy=signal.source_strategy,
            price_at_signal=price_at_signal,
            price_at_eval=price_at_eval,
            price_change_pct=round(price_change_pct, 4),
            direction_correct=direction_correct,
            pnl=pnl,
            evaluation=evaluation,
        )

        session.add(outcome)
        await session.commit()
        await session.refresh(outcome)

    return {
        "id": str(outcome.id),
        "symbol": outcome.symbol,
        "signal_type": outcome.signal_type,
        "direction_correct": direction_correct,
        "price_change_pct": outcome.price_change_pct,
        "evaluation": evaluation[:200],
    }


async def record_lesson(
    lesson_type: str,
    category: str,
    description: str,
    symbol: str | None = None,
    source_strategy: str | None = None,
    evidence: dict | None = None,
    ttl_days: int = 30,
) -> dict:
    """学びを記録する。類似の既存 lesson があれば observation_count を増やす

    Args:
        lesson_type: "signal_accuracy", "market_pattern", "risk", "strategy", "info_source"
        category: "positive", "negative", "neutral"
        description: 学びの内容
        symbol: 特定銘柄に紐づく場合
        source_strategy: マーケットID (us, jp, eu, uk)
        evidence: 根拠データ (JSON)
        ttl_days: 有効期間（日数、デフォルト30日）
    """
    async with async_session() as session:
        # 類似の既存 lesson を検索
        stmt = (
            select(Lesson)
            .where(Lesson.lesson_type == lesson_type)
            .where(Lesson.is_active == True)  # noqa: E712
            .where(Lesson.description == description)
        )
        if symbol:
            stmt = stmt.where(Lesson.symbol == symbol.upper())

        result = await session.execute(stmt)
        existing = result.scalar_one_or_none()

        if existing:
            existing.observation_count += 1
            existing.confidence = min(1.0, existing.confidence + 0.1)
            existing.updated_at = datetime.utcnow()
            existing.expires_at = datetime.utcnow() + timedelta(days=ttl_days)
            if evidence:
                existing.evidence = evidence
            await session.commit()
            await session.refresh(existing)
            return {
                "id": str(existing.id),
                "action": "updated",
                "observation_count": existing.observation_count,
                "confidence": existing.confidence,
                "description": existing.description[:200],
            }

        lesson = Lesson(
            lesson_type=lesson_type,
            category=category,
            description=description,
            symbol=symbol.upper() if symbol else None,
            source_strategy=source_strategy,
            evidence=evidence or {},
            confidence=0.5,
            expires_at=datetime.utcnow() + timedelta(days=ttl_days),
        )
        session.add(lesson)
        await session.commit()
        await session.refresh(lesson)

    return {
        "id": str(lesson.id),
        "action": "created",
        "observation_count": 1,
        "confidence": lesson.confidence,
        "description": lesson.description[:200],
    }


async def get_relevant_lessons(
    symbol: str | None = None,
    source_strategy: str | None = None,
    lesson_type: str | None = None,
    limit: int = 20,
) -> dict:
    """関連する学びを取得する（Research Agent のプロンプトに注入用）

    Args:
        symbol: 特定銘柄に絞る場合
        source_strategy: マーケットID でフィルタ
        lesson_type: タイプでフィルタ
        limit: 取得件数
    """
    now = datetime.utcnow()
    async with async_session() as session:
        stmt = (
            select(Lesson)
            .where(Lesson.is_active == True)  # noqa: E712
            .where((Lesson.expires_at == None) | (col(Lesson.expires_at) > now))  # noqa: E711
            .order_by(Lesson.confidence.desc(), Lesson.observation_count.desc())
            .limit(limit)
        )
        if symbol:
            # 銘柄固有 + 汎用(symbol=None) の両方を取得
            stmt = stmt.where((Lesson.symbol == symbol.upper()) | (Lesson.symbol == None))  # noqa: E711
        if source_strategy:
            stmt = stmt.where(
                (Lesson.source_strategy == source_strategy) | (Lesson.source_strategy == None)  # noqa: E711
            )
        if lesson_type:
            stmt = stmt.where(Lesson.lesson_type == lesson_type)

        result = await session.execute(stmt)
        lessons = result.scalars().all()

    return {
        "lessons": [
            {
                "id": str(l.id),
                "lesson_type": l.lesson_type,
                "category": l.category,
                "symbol": l.symbol,
                "description": l.description,
                "confidence": l.confidence,
                "observation_count": l.observation_count,
                "source_strategy": l.source_strategy,
                "updated_at": l.updated_at.isoformat(),
            }
            for l in lessons
        ],
        "count": len(lessons),
    }


async def get_signal_accuracy(
    source_strategy: str | None = None,
    days: int = 30,
) -> dict:
    """シグナル精度のサマリーを取得する

    Args:
        source_strategy: マーケットID でフィルタ
        days: 過去何日分を集計するか
    """
    cutoff = datetime.utcnow() - timedelta(days=days)
    async with async_session() as session:
        stmt = select(SignalOutcome).where(SignalOutcome.evaluated_at > cutoff)
        if source_strategy:
            stmt = stmt.where(SignalOutcome.source_strategy == source_strategy)

        result = await session.execute(stmt)
        outcomes = result.scalars().all()

    if not outcomes:
        return {"message": "No signal outcomes found", "accuracy": {}}

    total = len(outcomes)
    correct = sum(1 for o in outcomes if o.direction_correct)
    by_type = {}
    for o in outcomes:
        t = o.signal_type
        if t not in by_type:
            by_type[t] = {"total": 0, "correct": 0}
        by_type[t]["total"] += 1
        if o.direction_correct:
            by_type[t]["correct"] += 1

    for t in by_type:
        by_type[t]["accuracy"] = round(by_type[t]["correct"] / by_type[t]["total"], 4) if by_type[t]["total"] > 0 else 0.0

    avg_confidence = sum(o.confidence for o in outcomes) / total
    avg_price_change = sum(abs(o.price_change_pct or 0) for o in outcomes) / total

    return {
        "period_days": days,
        "total_signals": total,
        "overall_accuracy": round(correct / total, 4),
        "avg_confidence": round(avg_confidence, 4),
        "avg_abs_price_change_pct": round(avg_price_change, 4),
        "by_type": by_type,
    }

File: packages/server/src/mcp-server/tools/feedback.ts
/** フィードバックループ: シグナル評価・学習ツール */

import { eq, and, gt, or, isNull, desc } from "drizzle-orm";
import { db } from "../../db/client.js";
import { signals, signalOutcomes, lessons } from "../../db/schema.js";

export async function evaluateSignal(
  signalId: string,
  priceAtSignal: number,
  priceAtEval: number,
  evaluation: string,
  pnl: number | null = null,
): Promise<Record<string, unknown>> {
  const [signal] = await db
    .select()
    .from(signals)
    .where(eq(signals.id, signalId));

  if (!signal) {
    return { error: `Signal ${signalId} not found` };
  }

  const priceChangePct =
    priceAtSignal > 0
      ? ((priceAtEval - priceAtSignal) / priceAtSignal) * 100
      : 0;

  let directionCorrect: boolean;
  if (signal.signalType === "buy") {
    directionCorrect = priceChangePct > 0;
  } else if (signal.signalType === "sell") {
    directionCorrect = priceChangePct < 0;
  } else if (signal.signalType === "avoid") {
    directionCorrect = priceChangePct < 0;
  } else {
    // hold
    directionCorrect = Math.abs(priceChangePct) < 2.0;
  }

  const [outcome] = await db
    .insert(signalOutcomes)
    .values({
      signalId: signal.id,
      symbol: signal.symbol,
      signalType: signal.signalType,
      strength: signal.strength,
      confidence: signal.confidence ?? 0.5,
      sourceStrategy: signal.sourceStrategy ?? "",
      priceAtSignal,
      priceAtEval,
      priceChangePct: Math.round(priceChangePct * 10000) / 10000,
      directionCorrect,
      pnl,
      evaluation,
    })
    .returning();

  return {
    id: outcome.id,
    symbol: outcome.symbol,
    signal_type: outcome.signalType,
    direction_correct: directionCorrect,
    price_change_pct: outcome.priceChangePct,
    evaluation: evaluation.slice(0, 200),
  };
}

export async function recordLesson(
  lessonType: string,
  category: string,
  description: string,
  symbol: string | null = null,
  sourceStrategy: string | null = null,
  evidence: Record<string, unknown> | null = null,
  ttlDays = 30,
): Promise<Record<string, unknown>> {
  // 類似の既存 lesson を検索
  const conditions = [
    eq(lessons.lessonType, lessonType),
    eq(lessons.isActive, true),
    eq(lessons.description, description),
  ];
  if (symbol) {
    conditions.push(eq(lessons.symbol, symbol.toUpperCase()));
  }

  const [existing] = await db
    .select()
    .from(lessons)
    .where(and(...conditions));

  if (existing) {
    const newConfidence = Math.min(1.0, (existing.confidence ?? 0.5) + 0.1);
    const newExpires = new Date(Date.now() + ttlDays * 86400000);

    await db
      .update(lessons)
      .set({
        observationCount: (existing.observationCount ?? 1) + 1,
        confidence: newConfidence,
        updatedAt: new Date(),
        expiresAt: newExpires,
        ...(evidence ? { evidence } : {}),
      })
      .where(eq(lessons.id, existing.id));

    return {
      id: existing.id,
      action: "updated",
      observation_count: (existing.observationCount ?? 1) + 1,
      confidence: newConfidence,
      description: existing.description.slice(0, 200),
    };
  }

  const [lesson] = await db
    .insert(lessons)
    .values({
      lessonType,
      category,
      description,
      symbol: symbol?.toUpperCase() ?? null,
      sourceStrategy,
      evidence: evidence ?? {},
      confidence: 0.5,
      expiresAt: new Date(Date.now() + ttlDays * 86400000),
    })
    .returning();

  return {
    id: lesson.id,
    action: "created",
    observation_count: 1,
    confidence: lesson.confidence,
    description: lesson.description.slice(0, 200),
  };
}

export async function getRelevantLessons(
  symbol: string | null = null,
  sourceStrategy: string | null = null,
  lessonType: string | null = null,
  limit = 20,
): Promise<Record<string, unknown>> {
  const now = new Date();
  const conditions = [
    eq(lessons.isActive, true),
    or(isNull(lessons.expiresAt), gt(lessons.expiresAt, now)),
  ];

  if (symbol) {
    conditions.push(
      or(eq(lessons.symbol, symbol.toUpperCase()), isNull(lessons.symbol)),
    );
  }
  if (sourceStrategy) {
    conditions.push(
      or(
        eq(lessons.sourceStrategy, sourceStrategy),
        isNull(lessons.sourceStrategy),
      ),
    );
  }
  if (lessonType) {
    conditions.push(eq(lessons.lessonType, lessonType));
  }

  const rows = await db
    .select()
    .from(lessons)
    .where(and(...conditions))
    .orderBy(desc(lessons.confidence), desc(lessons.observationCount))
    .limit(limit);

  return {
    lessons: rows.map((l) => ({
      id: l.id,
      lesson_type: l.lessonType,
      category: l.category,
      symbol: l.symbol,
      description: l.description,
      confidence: l.confidence,
      observation_count: l.observationCount,
      source_strategy: l.sourceStrategy,
      updated_at: l.updatedAt?.toISOString(),
    })),
    count: rows.length,
  };
}

export async function getSignalAccuracy(
  sourceStrategy: string | null = null,
  days = 30,
): Promise<Record<string, unknown>> {
  const cutoff = new Date(Date.now() - days * 86400000);

  const conditions = [gt(signalOutcomes.evaluatedAt, cutoff)];
  if (sourceStrategy) {
    conditions.push(eq(signalOutcomes.sourceStrategy, sourceStrategy));
  }

  const rows = await db
    .select()
    .from(signalOutcomes)
    .where(and(...conditions));

  if (rows.length === 0) {
    return { message: "No signal outcomes found", accuracy: {} };
  }

  const total = rows.length;
  const correct = rows.filter((o) => o.directionCorrect).length;

  const byType: Record<
    string,
    { total: number; correct: number; accuracy: number }
  > = {};
  for (const o of rows) {
    const t = o.signalType;
    if (!byType[t]) byType[t] = { total: 0, correct: 0, accuracy: 0 };
    byType[t].total++;
    if (o.directionCorrect) byType[t].correct++;
  }
  for (const t of Object.keys(byType)) {
    byType[t].accuracy =
      byType[t].total > 0
        ? Math.round((byType[t].correct / byType[t].total) * 10000) / 10000
        : 0;
  }

  const avgConfidence =
    rows.reduce((sum, o) => sum + o.confidence, 0) / total;
  const avgPriceChange =
    rows.reduce((sum, o) => sum + Math.abs(o.priceChangePct ?? 0), 0) / total;

  return {
    period_days: days,
    total_signals: total,
    overall_accuracy: Math.round((correct / total) * 10000) / 10000,
    avg_confidence: Math.round(avgConfidence * 10000) / 10000,
    avg_abs_price_change_pct: Math.round(avgPriceChange * 10000) / 10000,
    by_type: byType,
  };
}
